apps/server/test/lexical/server/project/progress/state_test.exs

```diff
@@ -0,0 +1,53 @@
+defmodule Lexical.Server.Project.Progress.StateTest do
+  alias Lexical.Server.Project.Progress.State
+  alias Lexical.Server.Project.Progress.Value
+  alias Lexical.Server.Transport
+
+  import Lexical.RemoteControl.Api.Messages
+  import Lexical.Test.Fixtures
+
+  use ExUnit.Case
+  use Patch
+
+  def with_a_project(_) do
+    project = project()
+    {:ok, %{project: project}}
+  end
+
+  setup :with_a_project
+  setup do: patch(Transport, :write, :ok)
+
+  def message(label, message \\ nil) do
+    project_progress(label: label, message: message)
+  end
+
+  test "it should be able to add a begin event and put the new token", %{project: project} do
+    label = "mix deps.get"
+    state = State.new(project) |> State.begin(message(label))
+
+    assert %Value{token: token, title: ^label, kind: :begin} = state.progress_by_label[label]
+    assert token != nil
+  end
+
+  test "it should be able to add a report event use the begin event token", %{project: project} do
+    label = "mix compile"
+    state = State.new(project) |> State.begin(message(label))
+
+    previous_token = state.progress_by_label[label].token
+
+    %{progress_by_label: progress_by_label} =
+      State.report(state, message(label, "lib/my_module.ex"))
+
+    assert %Value{token: ^previous_token, message: "lib/my_module.ex", kind: :report} =
+             progress_by_label[label]
+  end
+
+  test "clear the token_by_label after received a complete event", %{project: project} do
+    state = State.new(project) |> State.begin(message("mix compile"))
+
+    %{progress_by_label: progress_by_label} =
+      State.complete(state, message("mix compile", "in 2s"))
+
+    assert progress_by_label == %{}
+  end
```

Comment:
the happy path is well tested, but what about these scenarios:

1. A report message is received before a begin
2. An exception happens during a report message (Patch is great for this)
3. A complete message happens before a report

Reply:
I added tests for 1 and 3, and for `2`, I don't know how to test it.

And do you mean to test the `with_progress` logic? That is too simple, I don't think we need to test the `try...after` keyword.

```elixir
    try do
      send(progress_pid, project_progress(label: label, stage: :begin))
      func.()
    after
      send(progress_pid, project_progress(label: label, stage: :complete))
    end
```

Reply:
why don't you need to test the try/after? You're assuming that the code exists, you should test functionality, not implementation. 

Someone could go in tomorrow and change it somehow so it'd break, without a test, how would we know?

Reply:
makes sense.